src/ets/scala/io/iohk/ethereum/ets/blockchain/ScenarioSetup.scala

```diff
@@ -0,0 +1,143 @@
+import io.iohk.ethereum.vm.VM
+import org.spongycastle.util.encoders.Hex
+
+import scala.annotation.tailrec
+import scala.util.{Failure, Success, Try}
+
+abstract class ScenarioSetup(scenario: BlockchainScenario)
+  extends EphemBlockchainTestSetup
+  with ValidatorsBuilder
+  with BlockchainConfigBuilder {
+
+  val emptyWorld = blockchain.getWorldStateProxy(-1, UInt256.Zero, None)
+
+  override lazy val blockchainConfig = buildBlockchainConfig(scenario.network)
+
+  val ledger = new LedgerImpl(VM, blockchain, blockchainConfig)
+
+  def loadGenesis(): Block = {
+    val genesisBlock = scenario.genesisRLP match {
+      case Some(rlp) =>
+        val block = rlp.toArray.toBlock
+        assert(block.header == scenario.genesisBlockHeader.toBlockHeader,
+          "decoded genesis block header did not match the expectation")
+        block
+
+      case None =>
+        Block(scenario.genesisBlockHeader.toBlockHeader, BlockBody(Nil, Nil))
+    }
+
+    blockchain.save(genesisBlock)
+    blockchain.save(genesisBlock.header.hash, Nil)
+    blockchain.save(genesisBlock.header.hash, genesisBlock.header.difficulty)
+    genesisBlock
+  }
+
+  def loadInitialWorld(): InMemoryWorldStateProxy = {
+    InMemoryWorldStateProxy.persistState(getWorldState(scenario.pre))
+  }
+
+  def getFinalWorld(): InMemoryWorldStateProxy = {
+    InMemoryWorldStateProxy.persistState(getWorldState(scenario.postState))
+  }
+
+  def getBestBlock(): Option[Block] = {
+    val bestBlockNumber = storagesInstance.storages.appStateStorage.getBestBlockNumber()
+    blockchain.getBlockByNumber(bestBlockNumber)
+  }
+
+  private def buildBlockchainConfig(network: String): BlockchainConfig =  network match {
+    case "EIP150"     => new Eip150Config
+    case "Frontier"   => new FrontierConfig
+    case "Homestead"  => new HomesteadConfig
+    case "FrontierToHomesteadAt5" => new FrontierToHomesteadAt5
+    case "HomesteadToEIP150At5" => new HomesteadToEIP150At5
+
+    // Some default config, test will fail or be canceled
+    case _ => new FrontierConfig
+  }
+
+  private def decode(s: String): Array[Byte] = {
+    val stripped = s.replaceFirst("^0x", "")
+    Hex.decode(stripped)
+  }
+
+  // During decoding we cant expect some failures especially in bcInvalidRlPTests.json
+  private def decodeBlock(s: String): Option[Block] = {
+    Try(decode(s).toBlock) match {
+      case Success(block) => Some(block)
+      case Failure(ex) => {ex.printStackTrace(); None}
+    }
+  }
+
+  private def isInvalidBlock(blockDef: BlockDef): Boolean = {
+    blockDef.blockHeader.isEmpty && blockDef.transactions.isEmpty && blockDef.uncleHeaders.isEmpty
+  }
+
+  private def getInvalid: List[BlockDef] = {
+    scenario.blocks.filter(isInvalidBlock)
+  }
+
+  private def getBlocks(blocks: List[BlockDef]): List[Block] = {
+    blocks.flatMap(blockDef => decodeBlock(blockDef.rlp))
+  }
+
+  def getAllBlocks: List[Block] = {
+    getBlocks(scenario.blocks)
+  }
+
+  def getInvalidBlocks: List[Block] = {
```

Comment:
Is it worth creating new functions for something that is used only once and is one-liner anyway?

Reply:
maybe, in that case, it is a little bit too much